File: slp/backend-dotnet/Features/Report/ReportService.cs
using backend_dotnet.Features.Admin;
using backend_dotnet.Features.Helpers;

namespace backend_dotnet.Features.Report;

public class ReportService : IReportService
{
    private readonly IReportRepository _reportRepo;
    private readonly IAdminLogRepository _adminLogRepo;

    public ReportService(IReportRepository reportRepo, IAdminLogRepository adminLogRepo)
    {
        _reportRepo = reportRepo;
        _adminLogRepo = adminLogRepo;
    }

    public async Task<ReportDto?> GetByIdAsync(int id)
    {
        var report = await _reportRepo.GetByIdAsync(id);
        return report == null ? null : MapToDto(report);
    }

    public async Task<PaginatedResult<ReportDto>> GetUnresolvedAsync(string? search = null, int page = 1, int pageSize = 20)
    {
        var (reports, total) = await _reportRepo.GetUnresolvedAsync(search, page, pageSize);
        var dtos = reports.Select(MapToDto).ToList();

        return new PaginatedResult<ReportDto>
        {
            Items = dtos,
            Total = total,
            Page = page,
            PageSize = pageSize
        };
    }

    public async Task<ReportDto> CreateAsync(int userId, CreateReportRequest request)
    {
        var report = new Report
        {
            UserId = userId,
            TargetType = request.TargetType,
            TargetId = request.TargetId,
            Reason = request.Reason,
            AttemptId = request.AttemptId,
            CreatedAt = DateTime.UtcNow
        };
        var created = await _reportRepo.CreateAsync(report);
        return MapToDto(created);
    }

    public async Task<bool> ResolveAsync(int adminId, int reportId)
    {
        var success = await _reportRepo.ResolveAsync(reportId, adminId);
        if (success)
        {
            await _adminLogRepo.LogAsync(new AdminLog
            {
                AdminId = adminId,
                Action = "resolve_report",
                TargetType = "report",
                TargetId = reportId
            });
        }
        return success;
    }

    public async Task<IEnumerable<ReportDto>> GetByUserIdAsync(int userId)
    {
        var reports = await _reportRepo.GetByUserIdAsync(userId);
        return reports.Select(MapToDto);
    }

    public async Task<bool> DeleteAsync(int userId, int reportId)
    {
        var report = await _reportRepo.GetByIdAsync(reportId);
        if (report == null) return false;
        if (report.UserId != userId) return false;
        if (report.Resolved) return false;

        return await _reportRepo.DeleteAsync(reportId);
    }

    private ReportDto MapToDto(Report r)
    {
        return new ReportDto
        {
            Id = r.Id,
            UserId = r.UserId,
            Username = r.User?.Username ?? "unknown",
            TargetType = r.TargetType,
            TargetId = r.TargetId,
            Reason = r.Reason,
            Resolved = r.Resolved,
            ResolvedAt = r.ResolvedAt,
            CreatedAt = r.CreatedAt,
            AttemptId = r.AttemptId
        };
    }
}

File: slp/backend-nestjs/src/modules/report/report.service.ts
import { Inject, Injectable } from "@nestjs/common";
import type { IReportRepository } from "./report.repository";
import { AdminLogRepository } from "../admin/admin-log.repository";
import { Report } from "./report.entity";
import { ReportDto } from "./dto/report.dto";
import { CreateReportRequest } from "./dto/create-report-request.dto";

export interface IReportService {
  getById(id: number): Promise<ReportDto | null>;
  getUnresolved(): Promise<ReportDto[]>;
  create(userId: number, request: CreateReportRequest): Promise<ReportDto>;
  resolve(adminId: number, reportId: number): Promise<boolean>;
  getByUserId(userId: number): Promise<ReportDto[]>;
  delete(userId: number, reportId: number): Promise<boolean>;
}

@Injectable()
export class ReportService implements IReportService {
  constructor(
    @Inject("IReportRepository") private readonly reportRepo: IReportRepository,
    private readonly adminLogRepo: AdminLogRepository,
  ) {}

  async getById(id: number): Promise<ReportDto | null> {
    const report = await this.reportRepo.findById(id);
    return report ? this.mapToDto(report) : null;
  }

  async getUnresolved(): Promise<ReportDto[]> {
    const reports = await this.reportRepo.findUnresolved();
    return reports.map((r) => this.mapToDto(r));
  }

  async create(
    userId: number,
    request: CreateReportRequest,
  ): Promise<ReportDto> {
    const report = new Report();
    report.userId = userId;
    report.targetType = request.targetType;
    report.targetId = request.targetId;
    report.reason = request.reason;
    report.attemptId = request.attemptId ?? null;
    report.createdAt = new Date();

    const created = await this.reportRepo.create(report);
    return this.mapToDto(created);
  }

  async resolve(adminId: number, reportId: number): Promise<boolean> {
    const success = await this.reportRepo.resolve(reportId, adminId);
    if (success) {
      await this.adminLogRepo.log({
        adminId,
        action: "resolve_report",
        targetType: "report",
        targetId: reportId,
      });
    }
    return success;
  }

  async getByUserId(userId: number): Promise<ReportDto[]> {
    const reports = await this.reportRepo.findByUserId(userId);
    return reports.map((r) => this.mapToDto(r));
  }

  async delete(userId: number, reportId: number): Promise<boolean> {
    const report = await this.reportRepo.findById(reportId);
    if (!report) return false;
    if (report.userId !== userId) return false;
    if (report.resolved) return false;

    return await this.reportRepo.delete(reportId);
  }

  private mapToDto(report: Report): ReportDto {
    return {
      id: report.id,
      userId: report.userId,
      username: report.user?.username ?? "unknown",
      targetType: report.targetType,
      targetId: report.targetId,
      reason: report.reason,
      resolved: report.resolved,
      resolvedAt: report.resolvedAt ?? undefined,
      createdAt: report.createdAt,
      attemptId: report.attemptId ?? undefined,
    };
  }
}
